### PR TITLE
Implement async module loading in CLI

### PR DESCRIPTION
**Pull request by ry**
Depends on #2082

Fixes https://github.com/denoland/deno/issues/1769

### CODE REVIEW

File: cli/main.rs [hunk 1]
```diff
@@ -102,17 +145,18 @@ fn main() {
...
     "main".to_string(),
     startup_data::deno_isolate_init(),
     state.clone(),
   );
 
-  let main_future = lazy(move || {
-    // Setup runtime.
-    js_check(main_worker.execute("denoMain()"));
+  // TODO(ry) somehow combine the two branches below. They're very similar but
+  // it's difficult to get the types to workout.
```
Comment (bartlomieju) [inline cli/main.rs]: Maybe split those conditional branches into separate functions? Then combined with clap's parsing we'd get:
```rust
  match matches.subcommand() {
    ("eval", Some(info_match)) => {
      eval_command(...);
    }
    ("info", Some(info_match)) => {
       info_command(...);
    }
    ("fmt", Some(fmt_match)) => {
       fmt_command(...);
    }
    (script, Some(script_match)) => {
      run_script_command();
    }
    _ => {}
  }
```
I can do that in follow-up PR

Comment (ry) [reply cli/main.rs]: That looks like a great way to organize it.

File: core/modules.rs [hunk 2]
```diff
@@ -45,135 +48,185 @@ pub trait Loader {
...
   }
 }
 
-impl<'l, L: Loader> Future for RecursiveLoad<'l, L> {
-  type Item = deno_mod;
-  type Error = Either<L::Error>;
+// TODO(ry) This is basically the same thing as RustOrJsError. They should be
+// combined into one type.
+#[derive(Debug, PartialEq)]
+pub enum Either<E> {
```
Comment (piscisaureus) [inline core/modules.rs]: This name is not great. EitherError would be better.

File: core/modules.rs [hunk 3]
```diff
@@ -45,135 +48,185 @@ pub trait Loader {
...
+  JSError(JSError),
+  Other(E),
+}
+
+impl<L: Loader> Future for RecursiveLoad<L> {
+  type Item = (deno_mod, L);
```
Comment (piscisaureus) [inline core/modules.rs]: I find that returning the loader from the future is really ugly and reeks. In the future we should try to refactor this.

File: core/modules.rs (hunk 2 above)
Comment (ry) [reply core/modules.rs]: Changed to JSErrorOr<E>

File: core/modules.rs (hunk 3 above)
Comment (ry) [reply core/modules.rs]: refactors welcome

**Comment by kevinkassimo** [comment]
@ry #2021 broken after this patch:
```
$ ./target/release/deno --reload info https://unpkg.com/@google/model-viewer@0.1.2/lib/model-viewer.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/model-viewer.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/features/animation.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/features/ar.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/features/controls.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/features/environment.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/features/loading.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/features/magic-leap.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/features/staging.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/model-viewer-base.js?module
Downloading https://unpkg.com/three@^0.101.0?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/ModelScene.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/SmoothControls.js?module
Downloading https://unpkg.com/lit-element@^2.0.0?module
Downloading https://unpkg.com/lit-element@^2.0.0/lib/updating-element?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/constants.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/template.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/Renderer.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/utils.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/Model.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/StaticShadow.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/ARRenderer.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/TextureUtils.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/WebGLUtils.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/assets/controls-svg.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/assets/view-in-ar-material-svg.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/third_party/three/EquirectangularToCubeGenerator.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/third_party/three/PMREMCubeUVPacker.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/third_party/three/PMREMGenerator.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/third_party/three/RGBELoader.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/EnvironmentMapGenerator.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/Reticle.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/CachingGLTFLoader.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/third_party/three/GLTFLoader.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/three-components/ModelUtils.js?module
Downloading https://unpkg.com/@google/model-viewer@0.1.2/lib/third_party/three/Sky.js?module
Downloading https://unpkg.com/lit-html@^1.0.0?module
Downloading https://unpkg.com/lit-html@^1.0.0/lib/shady-render.js?module
Downloading https://unpkg.com/lib/updating-element.js?module
Downloading https://unpkg.com/lib/decorators.js?module
Downloading https://unpkg.com/lit-html@^1.0.0/lit-html.js?module
Downloading https://unpkg.com/lib/css-tag.js?module
Import 'https://unpkg.com/lib@4.0.2/updating-element.js?module' failed: 404 Not Found
Import 'https://unpkg.com/lib@4.0.2/updating-element.js?module' failed: 404 Not Found
```

**Comment by ry** [comment]
@kevinkassimo ah cool thanks for the notice. I’ll try to get a unit test in core for it.

**Pull request merged**